llvm/lib/Target/EraVM/EraVMPostCodegenPrepare.cpp

```diff
@@ -326,11 +317,213 @@ bool EraVMPostCodegenPrepare::rearrangeOverflowHandlingBranches(Function &F) {
+static bool splitLoopPHILiveRanges(Function &F, LoopInfo &LI,
+                                   DominatorTree &DT) {
+  if (!EnableSplitLoopPHILiveRanges)
+    return false;
+
+  // This structure is used to store information about instructions in a block
+  // with their original immediate operand and the index of the last instruction
+  // that will be used to split the live range in the dominatee blocks.
+  struct InstSplitInfo {
+    using InstInfo = std::pair<Instruction *, APInt>;
+    SmallVector<InstInfo, 4> Insts;
+    uint64_t LastSplitIdx = UINT64_MAX;
+
+    // Add instruction and update the last split index if needed.
+    void addInst(Instruction *I) {
+      assert(isa<ConstantInt>(I->getOperand(1)) && "Expected constant operand");
+
+      if (LastSplitIdx == UINT64_MAX ||
+          !I->comesBefore(Insts[LastSplitIdx].first))
+        LastSplitIdx = Insts.size();
+
+      Insts.push_back({I, cast<ConstantInt>(I->getOperand(1))->getValue()});
+    }
+
+    // Get the last instruction that will be used to split the live range.
+    InstInfo &getLastSplitInst() {
+      assert(LastSplitIdx != UINT64_MAX && "Expected valid index");
+      return Insts[LastSplitIdx];
+    }
+  };
+
+  bool Changed = false;
+  for (auto *L : LI) {
+    for (auto &PN : L->getHeader()->phis()) {
+      if (PN.getNumUses() <= NumOfPHIUsesToSplitLiveRanges)
+        continue;
+
+      DomTreeNode *RootNode = DT.getNode(L->getHeader());
+      if (!RootNode)
+        continue;
+
+      MapVector<BasicBlock *, InstSplitInfo> Splits;
+      for (auto *U : PN.users()) {
+        // Only collect add instructions with constant operands in this loop.
+        auto *UI = cast<Instruction>(U);
+        if (!UI || UI->getOpcode() != Instruction::Add ||
+            !isa<ConstantInt>(UI->getOperand(1)) ||
+            LI.getLoopFor(UI->getParent()) != L)
+          continue;
+
+        auto [Itr, _] = Splits.insert({UI->getParent(), {}});
+        Itr->second.addInst(UI);
+      }
+
+      // Skip if there are no at least two blocks where we can split
+      // the live range.
+      if (Splits.size() < 2)
+        continue;
+
+      // Split ranges across blocks. This is done by finding the nearest
+      // dominator block, from which we can split the live range.
+      for (auto &[BB, Infos] : Splits) {
+        DomTreeNode *Node = DT.getNode(BB);
+        if (!Node)
+          continue;
+
+        // Find the nearest dominator, from which we can split the live range.
+        InstSplitInfo::InstInfo *NearestDominator = nullptr;
+        while ((Node = Node->getIDom())) {
+          auto *I = Splits.find(Node->getBlock());
+          if (I != Splits.end()) {
+            // We found the nearest dominator block, so take the last
+            // instruction from it.
+            NearestDominator = &I->second.getLastSplitInst();
+            break;
+          }
+
+          // Bail out if we reached the start of the loop.
+          if (Node == RootNode)
```

Comment:
We can't do this, since we could traverse over subloop BBs, like in this example:
```
bb1:
  dominator add
...
subloop bbs:
...
bb2:
  dominatee add
```
So to when we start searching from bb2 to find bb1, we could traverse over subloop bbs. Please note that BBs in the map are guaranteed to be in a loop, since there is a check for that.